_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,705 @@
+
+In the original `Msf::ModuleManager` that called `module_eval`, the lexical scope was `[Msf::ModuleManager, Msf]` because of the nested module declaration.
+
+![Nested module declarations for `Msf::ModuleManager`](/images/msf-module-manager-nested-modules.png)
+
+But, because I didn't want deeply indented code (or for the namespace modules to be defined in more than one file), I went with compact, unnested module declarations during the refactor:
+
+![Unnested module declarations for `Msf::Modules::Loader::Base`](/images/msf-modules-loader-base-unnested-modules.png)
+
+### Fixing constant look-up in `Msf::Modules::Loader::Base`
+
+I could have fixed the constant look-up by just changing to nested modules, but that would have indented all the code 3 more levels and made `Msf`, `Msf::Modules`, and `Msf::Modules::Loader` be defined in multiple files, and more likely to be loaded partially.  I also didn't think it wise that loaded Metasploit Modules were depending on the namespace (`Msf`) of the code loader, as this namespace could change since the project is called metasploit-framework and so by gem conventions should use the `Metasploit::Framework` namespace and not `Msf`.  This left me with the need to fake lexical scopes.  It turns out it's really hard to fake lexical scopes using just blocks, so I ended up having to use Strings with `module_eval` to construct the simulated lexical scope in a string and then call a method, `module_eval_with_lexical_scope` that was declared in that `String`, which calls `module_eval`.
+
+##### `namespace_modules_names`
+
+Escaping the `module_reference_name` was extracted to `namespace_module_names`.
+
+![`Msf::ModuleManager.wrapper_module` hex-encoding](/images/msf-module-manager-wrapper-module-hex-encoding.png) vs [![`Msf::Modules::Loader::Base#namespace_module_names`](/images/msf-modules-loader-base-namespace-module-names.png)](https://github.com/rapid7/metasploit-framework/blob/df9db42c32b572672ebd19fa4e6c7482c9876c2f/lib/msf/core/modules/loader/base.rb#L464-L495)
+
+After the escaping, `NAMESPACE_MODULE_NAMES` is prefixed.  `NAMESPACE_MODULE_NAMES` is `['Msf', 'Modules']`.  The `'Msf'` as the first element of the namespace is the key to reintroducing the `Msf` into the lexical scope.
+
+#### `create_namespace_module`
+
+[![`Msf::Modules::Loader::Base#create_namespace_module`](/images/msf-modules-loader-base-create-namespace-module.png)](https://github.com/rapid7/metasploit-framework/blob/df9db42c32b572672ebd19fa4e6c7482c9876c2f/lib/msf/core/modules/loader/base.rb#L328-L365)
+
+##### Nesting string Modules
+
+To build up a lexical scope, the code needs to wrap child modules in `module Parent ... end` blocks.  Instead of have to prepend `module Parent` to the front of a string and append `end` to the end of the a string, I reverse the `namespace_module_names` so I can build the inner `module`s first and then wrap them with the parent module like Russian nesting dolls.
```

Comment:
Instead of **~~have~~** **having** to prepend `module Parent` to the front of a string and append `end` to the end of the a string,